clipboard_unix.go

```diff
@@ -48,89 +43,98 @@ var (
 	missingCommands = errors.New("No clipboard utilities available. Please install xsel, xclip, wl-clipboard or Termux:API add-on for termux-clipboard-get/set.")
 )
 
-func init() {
+type commandInfo struct {
+	trimDOS bool
+
+	pasteCmdArgs []string
+	copyCmdArgs  []string
+
+	unsupported bool
+}
+
+func findClipboardUtility() commandInfo {
+	c := commandInfo{}
+
 	if os.Getenv("WAYLAND_DISPLAY") != "" {
-		pasteCmdArgs = wlpasteArgs
-		copyCmdArgs = wlcopyArgs
+		c.pasteCmdArgs = wlpasteArgs
+		c.copyCmdArgs = wlcopyArgs
 
 		if _, err := exec.LookPath(wlcopy); err == nil {
 			if _, err := exec.LookPath(wlpaste); err == nil {
-				return
+				return c
 			}
 		}
 	}
 
-	pasteCmdArgs = xclipPasteArgs
-	copyCmdArgs = xclipCopyArgs
+	c.pasteCmdArgs = xclipPasteArgs
+	c.copyCmdArgs = xclipCopyArgs
 
 	if _, err := exec.LookPath(xclip); err == nil {
-		return
+		return c
 	}
 
-	pasteCmdArgs = xselPasteArgs
-	copyCmdArgs = xselCopyArgs
+	c.pasteCmdArgs = xselPasteArgs
+	c.copyCmdArgs = xselCopyArgs
 
 	if _, err := exec.LookPath(xsel); err == nil {
-		return
+		return c
 	}
 
-	pasteCmdArgs = termuxPasteArgs
-	copyCmdArgs = termuxCopyArgs
+	c.pasteCmdArgs = termuxPasteArgs
+	c.copyCmdArgs = termuxCopyArgs
 
 	if _, err := exec.LookPath(termuxClipboardSet); err == nil {
 		if _, err := exec.LookPath(termuxClipboardGet); err == nil {
-			return
+			return c
 		}
 	}
 
-	pasteCmdArgs = powershellExePasteArgs
-	copyCmdArgs = clipExeCopyArgs
-	trimDos = true
+	c.pasteCmdArgs = powershellExePasteArgs
+	c.copyCmdArgs = clipExeCopyArgs
+	c.trimDOS = true
 
 	if _, err := exec.LookPath(clipExe); err == nil {
 		if _, err := exec.LookPath(powershellExe); err == nil {
-			return
+			return c
 		}
 	}
 
-	Unsupported = true
+	return commandInfo{unsupported: true}
 }
 
-func getPasteCommand() *exec.Cmd {
-	if Primary {
-		pasteCmdArgs = pasteCmdArgs[:1]
-	}
-	return exec.Command(pasteCmdArgs[0], pasteCmdArgs[1:]...)
+func getPasteCommand(c commandInfo) *exec.Cmd {
+	return exec.Command(c.pasteCmdArgs[0], c.pasteCmdArgs[1:]...)
 }
 
-func getCopyCommand() *exec.Cmd {
-	if Primary {
-		copyCmdArgs = copyCmdArgs[:1]
-	}
-	return exec.Command(copyCmdArgs[0], copyCmdArgs[1:]...)
+func getCopyCommand(c commandInfo) *exec.Cmd {
+	return exec.Command(c.copyCmdArgs[0], c.copyCmdArgs[1:]...)
 }
 
 func readAll() (string, error) {
-	if Unsupported {
+	c := findClipboardUtility()
```

Comment:
Hi :wave: just a small comment: instead of instantiating a new `commandInfo` every time `readAll()` or `writeAll()` is called, should it be possible to create a singleton:

```go
 var instance *commandInfo
 var once sync.Once

 func GetInstance() *commandInfo {
     once.Do(func() {
         instance = findClipboardUtility()
     })
     return instance
 }
```

And call `c := GetInstance()` instead of `c := findClipboardUtility()`?

It will be quicker since `findClipboardUtility()` will be called only once (at the first call).

What do you think? BTW, thanks for this PR :raised_hands:

Reply:
I considered this option, but it didn't seem worth the additional complication for a call that is unlikely to be called very often. If making this change helps the PR get merged though, I'm all for it.